nginx/lua/gossip.lua

```diff
@@ -0,0 +1,132 @@
+    for peer in string.gmatch(peerstr, "[^,]+") do
+        peers[peer] = true
+    end
+    return peers, starting
+end
+
+---@type function
+---@param peer string
+---@return PeerData peerdata
+function Gossip.get_peerdata(peer)
+    local epoch = ngx.shared.gossip_data:get("epoch:" .. peer)
+    if not epoch then
+        return {epoch = -1, status = "unknown", timestamp = 0, server_version = "unknown"}
+    end
+    local status = ngx.shared.gossip_data:get("status:" .. peer)
+    local timestamp = ngx.shared.gossip_data:get("timestamp:" .. peer)
+    local server_version = ngx.shared.gossip_data:get("server_version:" .. peer)
+    return {epoch = epoch, status = status, timestamp = timestamp, server_version = server_version}
+end
+
+---@type function
+---@param peer string
+---@param peerdata PeerData
+function Gossip.set_peerdata(peer, peerdata)
+    ngx.shared.gossip_data:set("epoch:" .. peer, peerdata.epoch)
+    ngx.shared.gossip_data:set("status:" .. peer, peerdata.status)
+    ngx.shared.gossip_data:set("timestamp:" .. peer, peerdata.timestamp)
+    ngx.shared.gossip_data:set("server_version:" .. peer, peerdata.server_version)
+end
+
+---@type function
+---@param peer string
+---@param peerdata PeerData
+function Gossip.update_peerdata(peer, peerdata)
+    local current_timestamp = ngx.shared.gossip_data:get("timestamp:" .. peer)
+    if current_timestamp and current_timestamp > peerdata.timestamp then
+        return
+    elseif current_timestamp == nil and peer ~= config.data.server_address then
+        -- Check if we agree on http or https and if not we don't add the peer
+        if peer:sub(0, 5) ~= config.data.server_address:sub(0, 5) then
+            ngx.log(ngx.WARN, "Not adding new peer " .. peer .. " because it does not have the same security level as us")
+            return
+        end
+        ngx.log(ngx.WARN, "Adding new peer " .. peer)
+        -- Set its data first to reduce race conditions?
+        Gossip.set_peerdata(peer, peerdata)
+        -- Potential TODO: factor this out to a once-per-gossip update
+        local peerstr = ngx.shared.gossip_data:get("peers")
+        ---@cast peerstr string
+        if peerstr:find(peer, 0, true) == nil then
+            ngx.shared.gossip_data:set("peers", peerstr .. "," .. peer)
+        end
+    else
+        Gossip.set_peerdata(peer, peerdata)
+    end
+end
+
+
+---@type function
+---@return string message
+---Prepare a message to send to peers representing our current view of the network
+function Gossip.prepare_message()
+    local peers, _ = Gossip.peers()
+
+    local message = {}
+    for peer,_ in pairs(peers) do
+        local peerdata = Gossip.get_peerdata(peer)
+        table.insert(message, {
+            name = peer,
+            data = peerdata,
+        })
+    end
+
+    -- Update our own data
+    local selfdata = Gossip.get_peerdata(config.data.server_address)
+    selfdata.epoch = selfdata.epoch + 1
+    selfdata.status = "alive"
```

Comment:
In the future we could set more useful status messages, e.g. how healthy we think we are, etc.